datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,332 @@
+                print("Created index if not exists when nrows = %d" % nrows)
+                self.index_counter = 0
+
+        cur.close()
+        self.pool.putconn(conn)
+
+        return doc_ids
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        """
+        Takes in a list of queries with embeddings and filters and
+        returns a list of query results with matching document chunks and scores.
+        """
+        # Prepare query responses and results object
+        results: List[QueryResult] = []
+
+        # Create a connection
+        conn = self.pool.getconn()
+        cur = conn.cursor()
+
+        for query in queries:
+            query_results: List[DocumentChunkWithScore] = []
+            embedding = ",".join(str(v) for v in query.embedding)
+            query_statement = (
+                "SELECT doc_id, text, metadata, 1 - (%s.embedding <=> ARRAY[%s]::vector) AS score FROM %s ORDER BY score DESC LIMIT %d;"
+                % (
+                    POSTGRES_TABLENAME,
+                    embedding,
+                    POSTGRES_TABLENAME,
+                    query.top_k,
+                )
+            )
+            cur.execute(query_statement)
+            sql_query_results = cur.fetchall()
+
+            query_results: List[DocumentChunkWithScore] = []
+            for result in sql_query_results:
+                doc_id = result[0]
+                text = result[1]
+                metadata = result[2]
+                score = result[3]
+                query_result = DocumentChunkWithScore(
+                    id=doc_id, score=score, text=text, metadata=(metadata)
+                )
+                query_results.append(query_result)
+
+            results.append(QueryResult(query=query.query, results=query_results))
+
+        conn.commit()
+        cur.close()
+        self.pool.putconn(conn)
+
+        return results
+
+    async def delete(
+        self,
+        ids: List[str] | None = None,
+        filter: DocumentMetadataFilter | None = None,
+        delete_all: bool | None = None,
+    ) -> bool:
+        # Create a connection
+        conn = self.pool.getconn()
+        cur = conn.cursor()
+
+        if delete_all:
+            delete_statement = "TRUNCATE %s" % (POSTGRES_TABLENAME)
+            try:
+                cur.execute(delete_statement)
+            except Exception as e:
+                print(e)
+
+        if filter:
+            start_date = None
+            end_date = None
+
+            if "start_date" in filter.dict().keys():
+                start_date = filter.dict().pop("start_date")
+                if start_date:
+                    start_date = to_unix_timestamp(start_date)
+
+            if "end_date" in filter.dict().keys():
+                end_date = filter.dict().pop("end_date")
+                if end_date:
+                    end_date = to_unix_timestamp(end_date)
+
+            if start_date and end_date:
```

Comment:
Since these columns act as filters in a predicate, it will significantly improve deletions if you explicitly index them in the JSONB metadata, or move them out into structured columns with indexes.